plugin/evm/block.go

```diff
@@ -102,17 +155,84 @@ func (b *Block) syntacticVerify() error {
 
 // Verify implements the snowman.Block interface
 func (b *Block) Verify(context.Context) error {
-	return b.verify(true)
+	return b.verify(nil, true)
+}
+
+// ShouldVerifyWithContext implements the block.WithVerifyContext interface
+func (b *Block) ShouldVerifyWithContext(context.Context) (bool, error) {
+	precompileConfigs := b.vm.chainConfig.AvalancheRules(b.ethBlock.Number(), b.ethBlock.Timestamp()).ActivePrecompiles
+
+	// Check if any of the transactions in the block list a precompile address in the access list.
+	for _, tx := range b.ethBlock.Transactions() {
+		for _, accessTuple := range tx.AccessList() {
+			if _, ok := precompileConfigs[accessTuple.Address]; ok {
+				log.Debug("Block verification requires proposerVM context", "block", b.ID(), "height", b.Height())
+				return true, nil
+			}
+		}
+	}
+
+	log.Debug("Block verification does not require proposerVM context", "block", b.ID(), "height", b.Height())
+	return false, nil
 }
 
-func (b *Block) verify(writes bool) error {
+// VerifyWithContext implements the block.WithVerifyContext interface
+func (b *Block) VerifyWithContext(ctx context.Context, proposerVMBlockCtx *block.Context) error {
+	if proposerVMBlockCtx != nil {
+		log.Debug("Verifying block with context", "block", b.ID(), "height", b.Height())
+	} else {
+		log.Debug("Verifying block without context", "block", b.ID(), "height", b.Height())
+	}
+
+	return b.verify(proposerVMBlockCtx, true)
+}
+
+// Verify the block is valid.
+// Enforces that the predicates are valid within [proposerVMBlockCtx].
+// Writes the block details to disk and the state to the trie manager iff writes=true.
+func (b *Block) verify(proposerVMBlockCtx *block.Context, writes bool) error {
 	if err := b.syntacticVerify(); err != nil {
 		return fmt.Errorf("syntactic block verification failed: %w", err)
 	}
 
+	// Only enforce predicates if the chain has already bootstrapped.
+	// If the chain is still bootstrapping, we can assume that all blocks we are verifying have
+	// been accepted by the network (so the predicate was validated by the network when the
+	// block was originally verified).
+	if b.vm.bootstrapped {
+		if err := b.verifyPredicates(proposerVMBlockCtx); err != nil {
+			return fmt.Errorf("failed to verify predicates: %w", err)
+		}
+	}
+
+	// The engine may call VerifyWithContext multiple times on the same block with different contexts.
+	// Since the engine will only call Accept/Reject once, we should only call InsertBlockManual once.
```

Comment:
How is it enforced to only call Accept/Reject once?

Reply:
This is an invariant maintained by the AvalancheGo consensus engine.

Reply:
See https://docs.avax.network/subnets/create-a-vm-blobvm#introduction for more details.

Reply:
See also the comments on the interface in AvalancheGo: https://github.com/ava-labs/avalanchego/blob/master/snow/consensus/snowman/block.go